### PR TITLE
[wasm-mt] A version of LowLevelLifoSemaphore that uses callbacks on the browser

### PR DESCRIPTION
**Pull request by lambdageek**
This is Part 2 of https://github.com/dotnet/runtime/issues/84489 - landing support for async JS interop on threadpool threads in multi-threaded WebAssembly.

This PR adds two things:

1. A new unmanaged `LifoSemaphoreAsyncWait` semaphore that can use Emscripten's ability to push C calls from one thread to another in order to implement a callback-based semaphore - when a thread wants to wait, it sets up a success callback and a timeout callback, and then can return to the JS event loop.  When the semaphore is released, Emscripten will trigger the callback to run on the waiting thread.  If the wait times out, the timeout callback will run.
2. A new flavor of the managed `LowLevelLifoSemaphore` that is not allowed to use the normal `Wait()` function, and instead needs to use the callback-based `PrepareAsyncWait()` function.

### CODE REVIEW

File: src/libraries/System.Private.CoreLib/src/System/Threading/LowLevelLifoSemaphore.cs
Comment (jkoritzinsky) [inline src/libraries/System.Private.CoreLib/src/System/Threading/LowLevelLifoSemaphore.cs:44]: Are there places where we use `LowLevelLifoSemaphore` that we're keeping on the old "call `Wait`" plan? If not, can we instead `#ifdef`-out the implementation of `Wait` or make it always throw `PlatformNotSupportedException` in this configuration instead of introducing two separate modes that make different sets of method calls valid?

Comment (lambdageek) [reply src/libraries/System.Private.CoreLib/src/System/Threading/LowLevelLifoSemaphore.cs:44]: We need both, I think. There are legitimate uses for the synchronous version on the browser.l, too. Just not in the threadpool worker interop scenario.

I could instead do something more OO-y with a common base class that handles the Counters logic. Wasn't sure if it was worthwhile.

Comment (jkoritzinsky) [reply src/libraries/System.Private.CoreLib/src/System/Threading/LowLevelLifoSemaphore.cs:44]: I'd prefer that if there isn't a performance issue from doing so. But I know that this is a high perf area, so if it's not possible that's okay.

Comment (lambdageek) [reply src/libraries/System.Private.CoreLib/src/System/Threading/LowLevelLifoSemaphore.cs]: Made a common base class `LowLevelLifoSemaphoreBase` and two subclasses `LowLevelLifoSemaphore` (same as before) and `LowLevelLifoAsyncWaitSemaphore` (with `PrepareAsyncWait` instead of `Wait`).

**Comment by lambdageek** [comment]
There's a weird, but consitent failure on some of the wasm lanes during the build:

```
npm ERR! Cannot read property '@rollup/plugin-typescript' of undefined
  
  npm ERR! A complete log of this run can be found in:
  npm ERR!     /home/cloudtest_azpcontainer/.npm/_logs/2023-04-08T20_46_38_593Z-debug.log
```

I'm not sure if it's something in this PR, or some kind of CI issue.

**Comment by jkoritzinsky** [comment]
I don't have a lot of familiarity with this portion of the Mono code base. I can take a look, but someone from Mono that works on WASM should definitely be the main reviewer for that component.

**Review by kouvel — COMMENTED** [review]
Just one question, otherwise LGTM. I only did a quick scan over the native parts, probably would be good to have someone familiar with WASM review as well.

File: src/mono/System.Private.CoreLib/src/System/Threading/LowLevelLifoAsyncWaitSemaphore.Browser.Threads.Mono.cs
Comment (kg) [inline src/mono/System.Private.CoreLib/src/System/Threading/LowLevelLifoAsyncWaitSemaphore.Browser.Threads.Mono.cs:33]: named constants for kind please

File: src/mono/System.Private.CoreLib/src/System/Threading/LowLevelLifoAsyncWaitSemaphore.Browser.Threads.Mono.cs
Comment (kg) [inline src/mono/System.Private.CoreLib/src/System/Threading/LowLevelLifoAsyncWaitSemaphore.Browser.Threads.Mono.cs:58]: should we at least check whether the wait can complete instantly before entering the loop? or is that not a scenario that's possible for the lifo semaphore unlike other synchronization primitives

File: src/mono/System.Private.CoreLib/src/System/Threading/LowLevelLifoAsyncWaitSemaphore.Browser.Threads.Mono.cs
Comment (kg) [inline src/mono/System.Private.CoreLib/src/System/Threading/LowLevelLifoAsyncWaitSemaphore.Browser.Threads.Mono.cs:105]: why is this here?

File: src/mono/System.Private.CoreLib/src/System/Threading/LowLevelLifoAsyncWaitSemaphore.Browser.Threads.Mono.cs
Comment (kg) [inline src/mono/System.Private.CoreLib/src/System/Threading/LowLevelLifoAsyncWaitSemaphore.Browser.Threads.Mono.cs:187]: yeah the timeoutms should be reduced, I think.

File: src/mono/System.Private.CoreLib/src/System/Threading/LowLevelLifoAsyncWaitSemaphore.Browser.Threads.Mono.cs
Comment (kg) [inline src/mono/System.Private.CoreLib/src/System/Threading/LowLevelLifoAsyncWaitSemaphore.Browser.Threads.Mono.cs:213]: According to the sample in https://learn.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.gchandle.fromintptr?view=net-8.0 I'm not sure it's correct to free the handle here. I think it might be a double-free, unless the caller isn't freeing their handle?

File: src/mono/System.Private.CoreLib/src/System/Threading/LowLevelLifoSemaphore.Unix.Mono.cs
Comment (kg) [inline src/mono/System.Private.CoreLib/src/System/Threading/LowLevelLifoSemaphore.Unix.Mono.cs:19]: named constant

File: src/mono/mono/utils/lifo-semaphore.c
Comment (kg) [inline src/mono/mono/utils/lifo-semaphore.c:26]: does this need to occur inside the lock?

File: src/mono/mono/utils/lifo-semaphore.c
Comment (kg) [inline src/mono/mono/utils/lifo-semaphore.c:239]: is this sufficient to ensure the lock is released before it runs? i assume it posts it to the message queue of that thread?

**Comment by kg** [comment]
I don't fully understand everything here, but other than my comments it looks OK

File: src/mono/System.Private.CoreLib/src/System/Threading/LowLevelLifoAsyncWaitSemaphore.Browser.Threads.Mono.cs
Comment (lambdageek) [reply src/mono/System.Private.CoreLib/src/System/Threading/LowLevelLifoAsyncWaitSemaphore.Browser.Threads.Mono.cs:33]: Actually I think I'm going to make two separate icalls.   `InitInternal(int kind)` is a holdover from before I had two separate subclasses of semaphore.  There's no good reason for it.

File: src/mono/System.Private.CoreLib/src/System/Threading/LowLevelLifoAsyncWaitSemaphore.Browser.Threads.Mono.cs
Comment (lambdageek) [reply src/mono/System.Private.CoreLib/src/System/Threading/LowLevelLifoAsyncWaitSemaphore.Browser.Threads.Mono.cs:105]: I was hoping someone would tell me that it's a good idea to keep the spin-waiting code. I'll delete it.

File: src/mono/System.Private.CoreLib/src/System/Threading/LowLevelLifoAsyncWaitSemaphore.Browser.Threads.Mono.cs
Comment (lambdageek) [reply src/mono/System.Private.CoreLib/src/System/Threading/LowLevelLifoAsyncWaitSemaphore.Browser.Threads.Mono.cs:187]: @kouvel are you familiar with this part of the (normal) LowLevelLifoSemaphore code: 

https://github.com/dotnet/runtime/blob/44ea019bbbaab1a0bb2877e6510036456a6495eb/src/libraries/System.Private.CoreLib/src/System/Threading/LowLevelLifoSemaphore.cs#L202-L204

The second time through the outer `while (true)`, if we lost the `InterlockedCompareExchange`, shouldn't `timeoutMs` be decreased by how long we have waited already?

Is it not done that way because measuring the elapsed waiting time is actually too expensive in the common case? (That could certainly be true on wasm, @kg since all the time stuff boils down to an external JS call to `performance.now`)

File: src/mono/System.Private.CoreLib/src/System/Threading/LowLevelLifoAsyncWaitSemaphore.Browser.Threads.Mono.cs
Comment (lambdageek) [reply src/mono/System.Private.CoreLib/src/System/Threading/LowLevelLifoAsyncWaitSemaphore.Browser.Threads.Mono.cs:213]: Yea this is different from the example in the docs.  In this case the ownership of the GC handle passes from `PrepareAsyncWaitCore` - which creates the GCHandle, to the underlying native wait structure, to the `SuccessCallback` or `TimeoutCallback`. At that point, the callback is the sole owner of the GC Handle (the underlying native wait entry was already freed).  So the callback is responsible for freeing the GCHandle.

In the example in the docs, the call to EnumWindows never owns the GCHandle - it is always owned by the calling function, which is responsible for freeing it.

File: src/mono/mono/utils/lifo-semaphore.c
Comment (lambdageek) [reply src/mono/mono/utils/lifo-semaphore.c:239]: yea the `_async` variant is like `queueMicrotask`: it will not run the callback immediately, so we will get a chance to unlock.

File: src/mono/System.Private.CoreLib/src/System/Threading/LowLevelLifoAsyncWaitSemaphore.Browser.Threads.Mono.cs
Comment (kouvel) [reply src/mono/System.Private.CoreLib/src/System/Threading/LowLevelLifoAsyncWaitSemaphore.Browser.Threads.Mono.cs:165]: You're right, the timeout should be reduced based on how long the wait took and looks like it's missing in the current implementation. I figure the overhead of timing would be much less than the overhead of WaitCore, though not sure about wasm.

File: src/mono/System.Private.CoreLib/src/System/Threading/LowLevelLifoAsyncWaitSemaphore.Browser.Threads.Mono.cs
Comment (kouvel) [reply src/mono/System.Private.CoreLib/src/System/Threading/LowLevelLifoAsyncWaitSemaphore.Browser.Threads.Mono.cs]: I guess whether to spin-wait or not would depend on perf and other considerations. For instance, is it better to yield back to the JS event loop immediately, or is it better and is it ok to spin-wait a bit for more work and only yield back after the spin-wait fails (potentially the spin-wait may succeed successively and not yield back to the JS event loop for a long time).

Comment (lambdageek) [reply src/mono/System.Private.CoreLib/src/System/Threading/LowLevelLifoAsyncWaitSemaphore.Browser.Threads.Mono.cs]: > potentially the spin-wait may succeed successively and not yield back to the JS event loop for a long time

I think this is a good reason to hold off on adding spin-waiting here until we see how real code behaves. Potentially starving JS wouldn't be great

Comment (kg) [reply src/mono/System.Private.CoreLib/src/System/Threading/LowLevelLifoAsyncWaitSemaphore.Browser.Threads.Mono.cs]: I think you can leave this code here for now but not enable it, because the risk of starving the main thread is big yeah

File: src/mono/System.Private.CoreLib/src/System/Threading/LowLevelLifoAsyncWaitSemaphore.Browser.Threads.Mono.cs
Comment (kg) [reply src/mono/System.Private.CoreLib/src/System/Threading/LowLevelLifoAsyncWaitSemaphore.Browser.Threads.Mono.cs:165]: performance.now is a little expensive, but it's still worth doing this right i think.

Comment (lambdageek) [reply src/mono/System.Private.CoreLib/src/System/Threading/LowLevelLifoAsyncWaitSemaphore.Browser.Threads.Mono.cs:165]: ok, I'll update the normal LowLevelLifoSemaphore and the async one to decrement `TimeoutMs`

File: src/libraries/System.Private.CoreLib/src/System/Threading/LowLevelLifoSemaphore.cs
Comment (kouvel) [inline src/libraries/System.Private.CoreLib/src/System/Threading/LowLevelLifoSemaphore.cs:139]: Could avoid timing if the timeout is -1

File: src/libraries/System.Private.CoreLib/src/System/Threading/LowLevelLifoSemaphore.cs
Comment (kouvel) [inline src/libraries/System.Private.CoreLib/src/System/Threading/LowLevelLifoSemaphore.cs:147]: This seems more like `endTicks` since it's an absolute-time tick value

File: src/libraries/System.Private.CoreLib/src/System/Threading/LowLevelLifoSemaphore.cs
Comment (kouvel) [inline src/libraries/System.Private.CoreLib/src/System/Threading/LowLevelLifoSemaphore.cs:181]: Maybe something like this instead would avoid a remote chance of `waitMs` being negative:

```c#
uint waitMs = (uint)(endTicks - startTicks);
if (waitMs < (uint)timeoutMs)
    timeoutMs -= (int)waitMs;
else
    timeoutMs = 0;
```

Also could update `if (!WaitCore(timeoutMs))` at the top of the loop to `if (timeoutMs == 0 || !WaitCore(timeoutMs))` to avoid waiting again with a zero timeout. Maybe similarly for wasm.

Comment (kouvel) [reply src/libraries/System.Private.CoreLib/src/System/Threading/LowLevelLifoSemaphore.cs:181]: Could just check for negative `waitMs` too, may be simpler:

```c#
int waitMs = endTicks - startTicks;
if (waitMs >= 0 && waitMs < timeoutMs)
    timeoutMs -= waitMs;
else
    timeoutMs = 0;
```

File: src/mono/System.Private.CoreLib/src/System/Threading/LowLevelLifoAsyncWaitSemaphore.Browser.Threads.Mono.cs
Comment (kouvel) [inline src/mono/System.Private.CoreLib/src/System/Threading/LowLevelLifoAsyncWaitSemaphore.Browser.Threads.Mono.cs:115]: It looks like `InternalAsyncWaitSuccess` calls `PrepareAsyncWaitCore` and not this method, so maybe the timeout check should be in `PrepareAsyncWaitCore`. There is an assert in this method above that the timeout here shouldn't be 0.

**Review by kouvel — COMMENTED** [review]
LGTM, thanks!

**Comment by lambdageek** [comment]
@kg @kouvel could I get a sign-off

**Review by kouvel — APPROVED** [review]

**Pull request merged**
